lib/tree.rb

```diff
@@ -16,51 +16,93 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(logn) 
+  # Space Complexity: O(n)
+  def add(key, value, current=@root)
+    # if the root is nil, the tree is empty, add a new node as root
+    return @root = TreeNode.new(key, value) unless @root
+
+    if key < current.key
+      return add(key, value, current.left) if current.left
+      current.left = TreeNode.new(key, value)
+    elsif key > current.key
+      return add(key, value, current.right) if current.right
+      current.right = TreeNode.new(key, value)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def find(key)
-    raise NotImplementedError
+  # Time Complexity: O(logn) 
+  # Space Complexity: O(n)
+  def find(key, current=@root)
+    return unless current
+
+    return current.value if current.key == key
+    return find(key, current.left) if key < current.key
+    return find(key, current.right) if key > current.key
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def inorder
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def inorder(current = @root, out = [])
+    return out unless current
+    
+    inorder(current.left, out)
+    out << { key: current.key, value: current.value }
+    inorder(current.right, out)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def preorder
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def preorder(current = @root, out = [])
+    return out unless current
+
+    out << { key: current.key, value: current.value }
+    preorder(current.left, out)
+    preorder(current.right, out)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def postorder
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def postorder(current = @root, out = [])
+    return out unless current
+
+    postorder(current.left, out)
+    postorder(current.right, out)
+    out << { key: current.key, value: current.value }
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def height
-    raise NotImplementedError
+  # Time Complexity: O(n) 
+  # Space Complexity: O(n)
+  def height(current = @root, h = 0)
```

Comment:
The space complexity is O(log n) if the tree is balanced and O(n) if it's not.
Otherwise this is great.  Nice work!